### ISSUE TITLE
web.http.Request.setHost ignores the port when setting the host header

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4918]
|<img alt="armooo's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| armooo reported|
|-|-|
|Trac ID|trac#4918|
|Type|defect|
|Created|2011-02-28 03:35:41Z|

While working on 3845 (which ended up looking like a dup of 2648) I noticed that web.http.Request.setHost is ignoring the port number when setting the host header.

Attachments:

* [.0001-web.http.Request.setHost-now-sets-the-port-in-the-ho.patch.swp](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bb4/bb4024c851d0022430375d6bbe33df9c2e053093/d3a950b8b4d14708cc1d891630685ccfa88a8e82.swp) (16384 bytes) - added by armooo on 2011-02-28 03:52:53Z - 
* [0001-web.http.Request.setHost-now-sets-the-port-in-the-ho.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bb4/bb4024c851d0022430375d6bbe33df9c2e053093/c3130743ff902a01852c208e938752358693d143.patch) (3778 bytes) - added by armooo on 2011-02-28 03:53:26Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4918 4918
type__defect defect
reporter__armooo armooo
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1298864141000000 1298864141000000
changetime__1299207836000000 1299207836000000
version__None None
owner__exarkun exarkun
cc__jknight cc__armooo
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="armooo's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|armooo removed owner|
|-|-|

Ignore the first attachment which is the wrong file.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to @exarkun|
|-|-|

Thank you for the patch, armooo.  This looks great.  One minor point, we're preferring `assertEquals` over `assertEqual` for new tests, and we use `camelCase` instead of `under_scores` for things like the `host_header` local in `setHost`.  Those are trivial though, I'll adjust the patch and then apply it.  Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set status to `closed`|
|-|-|

(In [30876]) Apply patch to cause Request.setHost to set a header which includes port information if appropriate.

Author: armooo
Reviewer: exarkun
Fixes: [#4918](https://github.com/twisted/twisted/issues/4918)

If the port passed to `Request.setHost` isn't the default for the scheme,
include it in the header added to `Request.requestHeaders` to more closely
imitate the header which would be present if the client were really issuing
requests to the specified host and port.
